lib/core.js

```diff
@@ -34,75 +34,98 @@ export function Core(cfg) {
+  this.sessionCookie = cfg.sessionCookie || 'ravelinSessionId';
   this.cookieExpiryDays = cfg.cookieExpiryDays || 365;
   this.cookies = new CookieJar({
     domain: cfg.cookieDomain
   });
 
-  if (cfg.id) {
-    this._id = this._idBackup(cfg.id);
-  } else {
-    this.sync();
-    if (cfg.init !== false) {
-      this.attach(cfg.syncMs || 2000);
-    }
+  this._id = resolve(this.Promise, cfg.id);
+  this.sync();
+  if (cfg.init !== false) {
+    this.attach(cfg.syncMs || 2000);
   }
 }
 
 /**
- * _idBackup wraps the user-given id, providing error reporting and a non-empty
- * fallback in-case it errors or provides an empty value.
- * @param {string|Promise<string>} id
- * @returns {Promise<string>}
+ * resolve returns a Promise that resolves to v. If v is already a promise, we
+ * just return it.
+ * @param {PromiseConstructor} Promise The promise implementation.
+ * @param {Promise|any} v The value we want in a then.
+ * @returns {Promise}
  */
-Core.prototype._idBackup = function(id) {
-  var that = this;
-  function backup() {
-    that._id = null;
-    that.sync();
-    return that.id();
+function resolve(Promise, v) {
+  if (v && typeof(v.then) === 'function') {
+    return v;
   }
-
-  // We could have handled id the same way - whether string or Promise - if new
-  // Promise(r => r(id)).then(sid => ...) always resolved to a string sid, but
-  // the Yaku Promise implementation we bundle doesn't unwrap the Promise id
-  // resolved in the constructor. So instead, we look for something then-able.
-  if (!id.then) {
-    return new this.Promise(function(r) { r(id); });
-  }
-
-  // Wrap the Promise id to ensure we never get an empty deviceId.
-  return this.sniffError(function() {
-    return id;
-  }).then(
-    function(id) {
-      if (id) return id;
-      return backup();
-    },
-    backup
-  );
-};
+  return new Promise(function(r) { r(v); });
+}
 
 /**
- * id reads our device identity from the browser.
- * @fulfil {string} The device ID.
+ * id reads our device ID from the browser.
  * @returns {Promise<string>}
  */
 Core.prototype.id = function() {
-  if (this._id) {
-    // Keep the IDs from last time.
-    return this._id;
+  return this.ids().then(function(ids) {
+    return ids.device;
+  });
+};
+
+/**
+ * @typedef {Object} IDs
+ * @prop {string} device
+ * @prop {string} session
+ */
+
+/**
+ * ids reads all IDs from the browser.
+ * @returns {Promise<IDs>}
+ */
+Core.prototype.ids = function() {
+  if (this._ids) {
+    return this._ids;
   }
 
   var c = this;
-  this._id = this.sniffError(function() {
-    return new c.Promise(function(resolve) {
-      resolve(c.cookies.get(c.cookie) || 'rjs-' + uuid());
+  this._ids =
+    this.sniffError(this._id)
+    .then(undefined, function() { /* swallow error. */ })
+    .then(function(d) {
+      var explicit = !!d;
+      if (!d) d = c.cookies.get(c.cookie);
+      var s = c.cookies.get(c.sessionCookie);
+
+      if (!explicit && s) {
```

Comment:
I suggest something like this instead of redefining the parameter. The `!!d` and then inverting it again `!explicit` could be a bit confusing too.

```suggestion
    .then(function(cfgId) {
      var d = cfgId || c.cookies.get(c.cookie);
      var s = c.cookies.get(c.sessionCookie);

      if (!cfgId && s) {
```

Reply:
Yes that's far nicer.